### ISSUE TITLE
Scalability issues with yume

### ISSUE THREAD

**Issue by dikim33** [oscar-cluster/oscar#258]
**Reported by bli on 31 Aug 2006 16:44 UTC**
I routinely use cexec to run yume commands on the cluster nodes for package installation, eg.:

`cexec 'yume -y install test'`

With a recent checkout, I always get error messages such as:

```
--------- x0065---------
ssh_exchange_identification: Connection closed by remote host
Usage: /usr/bin/yume OPTIONS [YUM_OPTIONS]

Call yum or repoquery with _only_ the repositories on the command line configured.
OPTIONS can be:
 --repo URL  : repository which should be included in yum actions
               (multiple --repo options are possible).
 --export    : create httpd configuration for local repositories
 --unexport  : delete httpd configuration for local repositories
 --listexp   : list exported local repositories
 --prepare   : generate repodata for local repositories
 --help      : this help information
 --repoquery : invoke repoquery command instead of yum
 --verbose   : verbose output

YUM_OPTIONS are options to be passed to the yum or repoquery command.
Setting the environment variable $YUME_VERBOSE to a number
between 1 and 10 will generate verbose output like multiple
-v options.
```

Since yume needs to ssh to the headnode to run distro-query, and perhaps the headnode's sshd got pounded by many simultaneous ssh sessions (there are 65 nodes in my cluster).

If we simply distribute distro-query as a package such that it can be installed on the client, can it eliminate the need to ssh to oscar_server?

**Comment by dikim33** [comment]
**Comment by bli on 7 Sep 2006 00:49 UTC**
I downloaded `yume` from r5332 and cpushed it to all the nodes, I am still getting the `ssh_exchange_identification` error and am also seeing another issue:

```
--------- x0006---------
Searching Packages:
Setting up repositories
not using ftp, http[or file for repos, skipping - /tftpboot/oscar/common-rpms
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
--------- x0007---------
Usage: /usr/bin/yume OPTIONS [YUM_OPTIONS](s],)

Call yum or repoquery with _only_ the repositories on the command line configured.
OPTIONS can be:
 --repo URL  : repository which should be included in yum actions
               (multiple --repo options are possible).
 --export    : create httpd configuration for local repositories
 --unexport  : delete httpd configuration for local repositories
 --listexp   : list exported local repositories
 --prepare   : generate repodata for local repositories
 --help      : this help information
 --repoquery : invoke repoquery command instead of yum
 --verbose   : verbose output

YUM_OPTIONS are options to be passed to the yum or repoquery command.
Setting the environment variable $YUME_VERBOSE to a number
between 1 and 10 will generate verbose output like multiple
-v options.

ssh_exchange_identification: Connection closed by remote host
```

Please let me know if I did something wrong here...

**Comment by dikim33** [comment]
**Comment by efocht on 7 Sep 2006 03:01 UTC**
I added some more debugging output to yume. Could you please try again? 
Thanks,
E.

**Comment by dikim33** [comment]
**Comment by bli on 7 Sep 2006 17:40 UTC**
New output of running `cexec 'yume search lam-oscar-modulefile'`:

```
--------- x0033---------
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
Searching Packages:
Setting up repositories
not using ftp, http[s], or file for repos, skipping - /tftpboot/oscar/common-rpms
--------- x0034---------
ssh_exchange_identification: Connection closed by remote host
Unexpected error detected (255) while distro-query. Exiting.
```

Note that `yume` worked prior to me patching it, but had ssh issues.

**Comment by dikim33** [comment]
**Comment by efocht on 11 Sep 2006 09:30 UTC**
This looks a bit like STDERR is not captured in the oscarinstall.log file.
Will check it and try to change it such that the messages logged are really useful...

**Comment by dikim33** [comment]
**Comment by efocht on 11 Sep 2006 14:18 UTC**
Hmmm, install_cluster seems to redirect STDERR correctly, and actually the test was outside the oscar installation, so I'm puzzled why the string "Connection closed by remote host" wasn't recognized on x0034.

I see no trace of retries or failed connections, this is strange. Bernard, are you sure cexec returns all STDERR messages? For debugging purposes you could try adding after line 418 a
   print "$_\n";
such that we get a chance to see why the match is missed. Don't know how to get closer to this otherwise.

**Comment by dikim33** [comment]
**Comment by bli on 12 Sep 2006 18:51 UTC**
I added the line:

```
                print "[$_\n";
```

after line 418 and now the output is:

```
--------- x0011---------
ssh_exchange_identification: Connection closed by remote host
Unexpected error detected (255) while distro-query. Exiting.
--------- x0012---------
[VARIABLE](VARIABLE]:): Distro package url : /tftpboot/distro/redhat-el-ws-4-i386
[OSCAR package pool : /tftpboot/oscar/common-rpms,/tftpboot/oscar/rhel-
4-i386
[VARIABLE](VARIABLE]:): distribution: redhat-el-ws
[distribution version: 4
[VARIABLE](VARIABLE]:): compat distribution: rhel
[compat distrover: 4
[VARIABLE](VARIABLE]:): packaging method: rpm
Cannot find a valid baseurl for repo: oscar_common-rpms
Searching Packages:
Setting up repositories
not using ftp, http[s], or file for repos, skipping - /tftpboot/oscar/common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
```

BTW, is the yume in pkgsrc trunk broken?  Because it is complaining that it could not find the repository for oscar_common-rpms.

**Comment by dikim33** [comment]
**Comment by efocht on 13 Sep 2006 10:34 UTC**
Bernard,

you must also update distro-query ! You seem to use an older variant of distro-query on the master node than expected by yume. The repositories returned to a node should be in the http://oscar_server/... format.

Ultimately, you can invoke yume with the -v option in order to see more details.
I'm still puzzled by the message you see from x0011. It seems that the error message is not coming from the ssh invoked by yume, but from the ssh invoked by cexec! Otherwise you would see a [VARIABLE] string in front of it. Can you imagine a reason why the cexec ssh would fail?

Regards,
Erich

**Comment by dikim33** [comment]
**Comment by efocht on 13 Sep 2006 10:38 UTC**
Wrong. You see the "Unexpected error detected (255) ...", so distro-query is running. In my experiments I saw this message appearing inside the <CMD> loop,
but maybe this comes through stderr. I'll look into a Perl manual again...

**Comment by dikim33** [comment]
**Comment by efocht on 13 Sep 2006 11:10 UTC**
Found that I need to do
open "$cmd 2>&1 |"
in order to have stderr in the same stream. This should solve the issue on x0011.
Checked into pkgsrc. Please try it.

**Comment by dikim33** [comment]
**Comment by bli on 13 Sep 2006 19:32 UTC**
I was using the latest version of `distro-query` from branch-5-0, which is from r5320:

```
# $Id: distro-query 5230 2006-07-27 14:33:39Z efocht $
```

However, just to be sure, I downloaded the latest `distro-query` code from trunk and got the same result.

Now yume re-tries the ssh connection until it works, so I think at least now you are guranteed that when you execute the yume command in parallel, they will run.  However I am still having issues with yume not working:

```
--------- x0004---------
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Searching Packages:
Setting up repositories
not using ftp, http[or file for repos, skipping - /tftpboot/oscar/common-rpms
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
--------- x0005---------
Warning: No xauth data; using fake authentication data for X11 forwarding.
Searching Packages:
Setting up repositories
not using ftp, http[s](s],), or file for repos, skipping - /tftpboot/oscar/common-rpms
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
```

I'll try to take a look at the code and see if I can figure this out.  Unfortunately this is a production server (with 65 nodes) and I cannot simply re-install OSCAR to re-test everything, so we need to do things in patches...

Thanks for your patience in this.

**Comment by dikim33** [comment]
**Comment by efocht on 13 Sep 2006 22:37 UTC**
The distro-query version I mean is r5248.

Try (manually) from a node (say x0011):

ssh oscar_server /opt/oscar\*/scripts/distro-query --node x0011

and check that the returned repositories are in a http:// format, and not just plain absolute paths. If they are only absolute paths, something is wrong. In routine print_image_detect() the absolute paths are rewritten to the exported repository names.

BTW: calling "distro-query --node x0011" should also work from the master node.

And: you only need to change distro-query and yume on your master. Nothing else should really matter. And distro-query is not used much in other places, AFAIR.

**Comment by dikim33** [comment]
**Comment by bli on 13 Sep 2006 22:52 UTC**
Hey Erich:

Yes, I'm using distro-query from trunk r5248.

Here's the output you requested:

```
[~](root@x0011)# ssh oscar_server /opt/oscar\*/scripts/distro-query --node x0011
Distro package url : http://oscar_server/repo/tftpboot/distro/redhat-el-ws-4-i386
OSCAR package pool : http://oscar_server/repo/tftpboot/oscar/common-rpms,http://oscar_server/repo/tftpboot/oscar/rhel-4-i386
distribution: redhat-el-ws
distribution version: 4
compat distribution: rhel
compat distrover: 4
packaging method: rpm
```

Seems that it's ok right?  What else could be wrong?

Here's the output if I run yume on x0011:

```
[~](root@x0011)# yume search blah
Searching Packages:
Setting up repositories
not using ftp, http[s], or file for repos, skipping - /tftpboot/oscar/common-rpms
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
```

Also, I only updated `distro-query` on the headnode, but for `yume` I usually cpush it to the nodes, because AFAIK it is needed on the clients too, right?

Thanks!

**Comment by dikim33** [comment]
**Comment by efocht on 14 Sep 2006 07:46 UTC**
Please try

yume -v search blah

from the client.

**Comment by dikim33** [comment]
**Comment by efocht on 14 Sep 2006 07:55 UTC**
On the client nodes you should not have $OSCAR_HOME set!
Also, you don't need distro-query there.
Regards,
Erich

**Comment by dikim33** [comment]
**Comment by bli on 14 Sep 2006 17:42 UTC**
Here's the output:

```
[~](root@x0002)# yume -v search blah
Executing: ssh oscar_server bash -l -c \\\$OSCAR_HOME/scripts/distro-query --node x0002
Repositories detected:
        /tftpboot/distro/redhat-el-ws-4-i386
        /tftpboot/oscar/common-rpms
        /tftpboot/oscar/rhel-4-i386
WARNING: repository name /tftpboot/oscar/common-rpms doesn't match any valid architecture
Repository architecture: i386
=== yum.conf file used (/tmp/filefSXLxE) ===
[package repository
baseurl=/tftpboot/distro/redhat-el-ws-4-i386

[oscar_common-rpms](main]
cachedir=/var/cache/yum
debuglevel=2
logfile=/var/log/yume.log
pkgpolicy=newest
#distroverpkg=redhat-release
tolerant=1
exactarch=1
retries=20
obsoletes=1
reposdir=
gpgcheck=0
[distro_redhat-el-ws-4-i386]
name=redhat-el-ws-4-i386)
name=common-rpms package repository
baseurl=/tftpboot/oscar/common-rpms

[package repository
baseurl=/tftpboot/oscar/rhel-4-i386

==========================================
Executing: yum -c /tmp/filefSXLxE search blah
Searching Packages:
Setting up repositories
not using ftp, http[s](oscar_rhel-4-i386]
name=rhel-4-i386), or file for repos, skipping - /tftpboot/oscar/common-rpms
Cannot find a valid baseurl for repo: oscar_common-rpms
Error: Cannot find a valid baseurl for repo: oscar_common-rpms
```

$OSCAR_HOME is not set on the clients, nor is `distro-query` there.

Looks like the `baseurl` is wrong...?  Thanks!

**Comment by dikim33** [comment]
**Comment by bli on 14 Sep 2006 18:14 UTC**
Okay I found out the likely cause of the problem.

This first output is using the command invoked by yume:

```
[~](root@x0002)# ssh oscar_server bash -l -c \\\$OSCAR_HOME/scripts/distro-query --node x0002.x330.bcgsc.bc.ca
Distro package url : /tftpboot/distro/redhat-el-ws-4-i386
OSCAR package pool : /tftpboot/oscar/common-rpms,/tftpboot/oscar/rhel-4-i386
distribution: redhat-el-ws
distribution version: 4
compat distribution: rhel
compat distrover: 4
packaging method: rpm
```

This output is from invoking the command which you told me to:

```
[~](root@x0002)# ssh oscar_server /opt/oscar\*/scripts/distro-query --node x0002
Distro package url : http://oscar_server/repo/tftpboot/distro/redhat-el-ws-4-i386
OSCAR package pool : http://oscar_server/repo/tftpboot/oscar/common-rpms,http://oscar_server/repo/tftpboot/oscar/rhel-4-i386
distribution: redhat-el-ws
distribution version: 4
compat distribution: rhel
compat distrover: 4
packaging method: rpm
```

Note that the one called via yume gave the incorrect URL, whereas the command you gave me provided the expected output.

Perhaps there is a bug in the command in the most recent yume...

**Comment by dikim33** [comment]
**Comment by efocht on 15 Sep 2006 10:28 UTC**
Bernard,

I thought this was clear. That's why I asked you to run the command manually.
As you can see, the command is different, yume calls
```
bash -l -c \\\$OSCAR_HOME/scripts/distro-query --node xyz
```
This (bash -l) was needed on SuSE in order to get the $OSCAR_HOME variable defined. Without "bash -l" the command fails on SuSE as $OSCAR_HOME is undefined. It seems that on RHEL this leads to a problem and I don't understand why. The fact that distro-query is executed at all is a good sign, it means that the escape backslashes are appropriate.

Please try the following:
```
ssh oscar_server bash -l -c \\\$OSCAR_HOME/scripts/distro-query --verbose --node x0002
```
and
```
ssh oscar_server bash -l -c \\\$OSCAR_HOME/scripts/distro-query --verbose --node x0002.x330.bcgsc.bc.ca
```
I hope we can learn something from this...

**Comment by dikim33** [comment]
**Comment by efocht on 15 Sep 2006 16:23 UTC**
Please try again with the latest yume in pkgsrc. Maybe it works now...
Will needd to test under SuSE, too.

**Comment by dikim33** [comment]
**Comment by bli on 15 Sep 2006 17:54 UTC**
Okay, this works fine under RHEL4, however, perhaps we can adjust the verbosity...

```
--------- x0035---------
Warning: No xauth data; using fake authentication data for X11 forwarding.
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Connection closed by remote host. Retrying 1/5
Searching Packages:
Setting up repositories
Reading repository metadata in from local files
No Matches found
```

Perhaps only show the "Connection closed by remote host" if we crank up verbosity.  Also, should the number "1/5" be incrementing?  What's the maximum retries or does it keep trying until it works?

Thanks for crushing this bug Erich.

**Comment by dikim33** [comment]
**Comment by efocht on 19 Sep 2006 21:28 UTC**
(In [5356]) Merged commit r5352 by efocht from 2006-09-19 10:51:34 +0200 (Tue, 19 Sep 2006)
with svnmerge_1by1 version v1.1 31.7.2006
from repository https://svn.oscar.openclustergroup.org/svn/oscar/trunk
to repository https://svn.oscar.openclustergroup.org/svn/oscar/branches/branch-5-0
Original commit message was:

Attempt to fix bugs #199 and #258.

Closes #199
Closes #258
